File: HSCTF2020/solve_aliens/solve_aliens.cpp
/*
 * solve_aliens.cpp
 *
 *  Created on: Jun 1, 2020
 *      Author: Stanley
 */

#include <iostream>
#include <fstream>

using namespace std;

typedef long long ll;

const int N = 500;
ll grid[N+1][N+1];
ll sums[N+1][N+1];
bool neg[N+1][N+1];

ll brute(int r1, int c1, int r2, int c2) {
	ll sum = 0;
	bool is_neg = false;
	for (int r=r1; r<=r2; ++r) {
		for (int c=c1; c<=c2; ++c) {
			sum += grid[r][c];
			if (grid[r][c] == -1) {
				is_neg = !is_neg;
			}
		}
	}


	if ((sum % 13) == 0) {
		if (is_neg) {
			sum = -sum;
		}
	}
	else {
		sum = 0;
	}

	cout << r1 << "," << c1 << " " << r2 << "," << c2 << " " << sum << " " << is_neg << endl;
	return sum;
}

int main() {
	ifstream fin("AlienMarking.txt");
	for (int i=1; i<=N; ++i) {
		for (int j=1; j<=N; ++j) {
			fin >> grid[i][j];
		}
	}

	for (int i=1; i<=N; ++i) {
		for (int j=1; j<=N; ++j) {
			sums[i][j] = grid[i][j] + sums[i-1][j] + sums[i][j-1] - sums[i-1][j-1];
			neg[i][j] = (grid[i][j] < 0) ^ neg[i-1][j] ^ neg[i][j-1] ^ neg[i-1][j-1];
//			cout << sums[i][j] << ' ';
		}
//		cout << endl;
	}

	ll ans = 0;
	for (int r1=1; r1<=N; ++r1) {
		cout << r1 << '\n';
		for (int c1=1; c1<=N; ++c1) {
			for (int r2=r1; r2<=N; ++r2) {
				for (int c2=c1; c2<=N; ++c2) {
//					ans += brute(r1,c1,r2,c2);
					ll cur = sums[r2][c2] - sums[r1-1][c2] - sums[r2][c1-1] + sums[r1-1][c1-1];
					bool is_neg = neg[r2][c2] ^ neg[r1 - 1][c2] ^ neg[r2][c1 - 1] ^ neg[r1-1][c1-1];

					if ((cur % 13) == 0) {
						if (is_neg) {
							cur = -cur;
						}
						ans += cur;
					}
					else {
						cur = 0;
					}
//					cout << r1 << "," << c1 << " " << r2 << "," << c2 << " " << cur << " " << is_neg << endl;
//					brute(r1, c1, r2, c2);
//					cout << endl;
				}
			}
		}
	}

	cout << ans << endl;
}

File: HSCTF2020/solve_aliens.py
def pprint(g: list):
    for r in g:
        print(r)
    print()


grid = []
with open("AlienMarking.txt", 'r') as inp:
    for line in inp:
        grid.append([int(n) for n in line.strip().split()])

# grid = [
#     [5, 7, 2, 6],
#     [4, 2, -1, 5],
#     [-1, 7, 6, 7],
#     [10, -1, -1, 8]
# ]

N = len(grid)
assert len(grid) == len(grid[0])

sums = [[grid[r][c] for c in range(N)] for r in range(N)]
negative = [[grid[r][c] == -1 for c in range(N)] for r in range(N)]
for r in range(N):
    for c in range(N):
        if r > 0:
            sums[r][c] += sums[r - 1][c]
            negative[r][c] ^= negative[r - 1][c]
        if c > 0:
            sums[r][c] += sums[r][c - 1]
            negative[r][c] ^= negative[r][c - 1]
        if r > 0 and c > 0:
            sums[r][c] -= sums[r - 1][c - 1]
            negative[r][c] ^= negative[r - 1][c - 1]

print("Prefix sum done")
ans = 0
for r1 in range(N):
    for c1 in range(N):
        print(f"Started {r1} {c1}")
        for r2 in range(r1, N):
            for c2 in range(c1, N):
                cur = sums[r2][c2]
                neg = negative[r2][c2]
                if r1 > 0 and c1 > 0:
                    cur -= sums[r1 - 1][c1 - 1]
                    neg ^= negative[r1 - 1][c1 - 1]
                elif r1 > 0:
                    cur -= sums[r1 - 1][c1]
                    neg ^= negative[r1 - 1][c1]
                elif c1 > 0:
                    cur -= sums[r1][c1 - 1]
                    neg ^= negative[r1][c1 - 1]

                if cur % 13 == 0:
                    if neg is True:
                        cur *= -1

                    ans += cur

# pprint(grid)
# pprint(sums)
# pprint(negative)
print(ans)
